### PR TITLE
IGNITE-15101 Fixes security context propagation for compute tasks.

### PR DESCRIPTION
**Pull request by ololo3000**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/clients/src/test/java/org/apache/ignite/internal/TaskEventSubjectIdSelfTest.java [hunk 1]
```diff
@@ -139,7 +139,7 @@ public void testSimpleTask() throws Exception {
         assert evt != null;
 
         assertEquals(EVT_TASK_STARTED, evt.type());
-        assertEquals(nodeId, evt.subjectId());
```
Comment (alex-plekhanov) [inline modules/clients/src/test/java/org/apache/ignite/internal/TaskEventSubjectIdSelfTest.java:142]: I don't get what this test is currently testing? Description says that it should test subject Id, but currently it checked that all subject ids is null. Perhaps we should enable security for this test and revert `evt.subjectId()` checking.

Comment (ololo3000) [reply modules/clients/src/test/java/org/apache/ignite/internal/TaskEventSubjectIdSelfTest.java:142]: I removed this test - the same test cases are included in the ComputeTaskRemoteSecurityContextTest, but it is more comprehensive.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteComputeImpl.java [hunk 2]
```diff
@@ -481,9 +481,10 @@ private IgniteComputeImpl(GridKernalContext ctx, ClusterGroupAdapter prj, UUID s
 
         try {
             ctx.task().setThreadContextIfNotNull(TC_SUBGRID_PREDICATE, prj.predicate());
-            ctx.task().setThreadContextIfNotNull(TC_SUBJ_ID, subjId);
 
-            return ctx.task().execute(taskName, arg, execName);
+            try (OperationSecurityContext ignored = ctx.security().withContext(subjId)) {
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteComputeImpl.java]: Removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/compute/ClientComputeTask.java [hunk 3]
```diff
@@ -100,15 +100,14 @@ void execute(long taskId, String taskName, Object arg, Set<UUID> nodeIds, byte f
...
-        UUID subjId = ctx.securityContext() == null ? null : ctx.securityContext().subject().id();
-
         task.setThreadContext(TC_SUBGRID_PREDICATE, nodePredicate);
-        task.setThreadContextIfNotNull(TC_SUBJ_ID, subjId);
         task.setThreadContext(TC_TIMEOUT, timeout);
         task.setThreadContext(TC_NO_FAILOVER, (flags & NO_FAILOVER_FLAG_MASK) != 0);
         task.setThreadContext(TC_NO_RESULT_CACHE, (flags & NO_RESULT_CACHE_FLAG_MASK) != 0);
 
-        taskFut = task.execute(taskName, arg);
+        try (OperationSecurityContext ignored = ctx.kernalContext().security().withContext(ctx.securityContext())) {
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/compute/ClientComputeTask.java]: Removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/security/IgniteSecurityProcessor.java [hunk 4]
```diff
@@ -367,13 +378,38 @@ else if (packAccess.contains(',' + IGNITE_INTERNAL_PACKAGE))
...
+    /** {@inheritDoc} */
+    @Override public void onLocalJoin() {
+        try {
+            SecurityContext secCtx = nodeSecurityContext(
+                marsh,
+                U.resolveClassLoader(ctx.config()),
+                ctx.discovery().localNode());
+
+            nodeSecCtxReadyFut.onDone(secCtx);
+        } catch (Throwable e) {
```
[the remark this answers is not in the document]
Comment (ololo3000) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/security/IgniteSecurityProcessor.java]: Done.

File: modules/clients/src/test/java/org/apache/ignite/common/ComputeTaskRemoteSecurityContextTest.java [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+
+        /** {@inheritDoc} */
+        @Override public ComputeJobResultPolicy result(ComputeJobResult res, List<ComputeJobResult> rcvd) {
+            if (res.getException() != null)
+                throw new IgniteException(res.getException());
+
+            return ComputeJobResultPolicy.WAIT;
+        }
+
+
```
Comment (alex-plekhanov) [inline modules/clients/src/test/java/org/apache/ignite/common/ComputeTaskRemoteSecurityContextTest.java:520]: Redundant NL

**Review by alex-plekhanov — APPROVED** [review]

**Pull request closed without merging**